server/app/controllers/applicant/UpsellController.java

```diff
@@ -165,4 +181,56 @@ public CompletionStage<Result> considerRegister(
               throw new RuntimeException(ex);
             });
   }
+
+  /** Download a PDF file of the application to the program. */
+  @Secure
+  public CompletionStage<Result> download(
+      Http.Request request, long programId, long applicationId, long applicantId)
+      throws ProgramNotFoundException {
+    if (!settingsManifest.getApplicationExportable(request)) {
+      return CompletableFuture.completedFuture(forbidden());
+    }
+    CompletableFuture<ProgramDefinition> program =
+        programService.getProgramDefinitionAsync(programId).toCompletableFuture();
+    Optional<CiviFormProfile> profileMaybe = profileUtils.currentUserProfile(request);
+    CiviFormProfile profile =
+        profileMaybe.orElseThrow(
+            () -> new NoSuchElementException("User authorized as applicant but no profile found"));
+
+    CompletableFuture<Account> account =
+        program
+            .thenComposeAsync(
+                v -> checkApplicantAuthorization(request, applicantId), httpContext.current())
+            .thenComposeAsync(v -> profile.getAccount(), httpContext.current())
+            .toCompletableFuture();
```

Comment:
**suggestion (blocking):** 

1. the `account` is not actually needed, it's only consumed by the `allOf` below but then its value is never read after that.
2. there's no reason to wait on the `program` completion stage resolving before making the call to `checkApplicantAuthorization`